### PR TITLE
Screenshot automation

### PR DESCRIPTION
**Pull request by pocmo**
Those are all patches. It's a lot. Looking at the single commits (and the titles) might be better than looking at everything at once. :)
This works locally in a docker container. Now I'm going to try this on taskcluster.

### CODE REVIEW

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+            } catch (IOException e) {
+                throw new AssertionError("Could not stop web server", e);
+            }
+        }
+    };
+
+    public static void swipeDownNotificationBar (UiDevice deviceInstance) {
+        int dHeight = deviceInstance.getDisplayHeight();
+        int dWidth = deviceInstance.getDisplayWidth();
+        int xScrollPosition = dWidth/2;
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+        int yScrollStop = dHeight/4 * 3;
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+        TestHelper.AboutItem.click();
+        TestHelper.webView.waitForExists(waitingTime);
+        Screengrab.screenshot("About_Page");
+        device.pressBack();
+    }
+
+    private void takeScreenshotOfHelpPage(UiDevice device) throws UiObjectNotFoundException {
+        TestHelper.menuButton.perform(click());
+        TestHelper.HelpItem.click();
+        TestHelper.webView.waitForExists(waitingTime*5);
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '\*' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '\*' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+        device.openNotification();
+        TestHelper.notificationBarDeleteItem.waitForExists(waitingTime);
+        Screengrab.screenshot("DeleteHistory_NotificationBar");
+        device.pressBack();
+    }
+
+    private void takeScreenshotOfEraseSnackbar(UiDevice device) {
+        TestHelper.webView.waitForExists(waitingTime);
+        TestHelper.floatingEraseButton.perform(click());
+        device.wait(Until.findObject(By.res("org.mozilla.focus.debug","snackbar_text")), waitingTime);
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: ',' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+
+        TestHelper.inlineAutocompleteEditText.waitForExists(waitingTime);
+        TestHelper.inlineAutocompleteEditText.clearTextField();
+        TestHelper.inlineAutocompleteEditText.setText(webServer.url(TEST_PATH).toString());
+        TestHelper.pressEnterKey();
+        TestHelper.webView.waitForExists(waitingTime);
+        titleMsg.waitForExists(waitingTime);
+        assertTrue("Website title loaded", titleMsg.exists());
+        Assert.assertTrue(mozillaImage.exists());
+        mozillaImage.dragTo(mozillaImage,7);
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: ',' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+    }
+
+    private void takeScreenshotOfErrorPages(UiDevice device) throws UiObjectNotFoundException {
+        UiObject tryAgainBtn = device.findObject(new UiSelector()
+                .resourceId("errorTryAgain")
+                .clickable(true));
+
+        for (ScreenGrabTest.ErrorTypes error: ScreenGrabTest.ErrorTypes.values()) {
+            TestHelper.inlineAutocompleteEditText.waitForExists(waitingTime);
+            TestHelper.inlineAutocompleteEditText.setText("error:"+ error.value);
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '\+' is not preceded with whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+import tools.fastlane.screengrab.locale.LocaleTestRule;
+
+import static android.support.test.espresso.action.ViewActions.click;
+import static android.view.KeyEvent.KEYCODE_ENTER;
+import static junit.framework.Assert.assertTrue;
+import static org.mozilla.focus.activity.TestHelper.browserURLbar;
+import static org.mozilla.focus.fragment.FirstrunFragment.FIRSTRUN_PREF;
+
+@RunWith(AndroidJUnit4.class)
+public class ScreenGrabTest {
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[PMD\] INFO: This class has too many methods, consider refactoring it\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+        //Open a new tab
+        openNewTabTitle.click();
+        TestHelper.webView.waitForExists(waitingTime);
+        multiTabBtn.click();
+        eraseHistoryBtn.waitForExists(waitingTime);
+        Screengrab.screenshot("Multi_Tab_Menu");
+        eraseHistoryBtn.click();
+    }
+
+    private void takeScreenshotOfGooglePlayDialog(UiDevice device) throws UiObjectNotFoundException {
```
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[PMD\] INFO: Avoid unused private methods such as 'takeScreenshotOfGooglePlayDialog\(UiDevice\)'\.

File: Screengrabfile [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+tests_apk_path 'app/build/outputs/apk/app-focus-webkit-debug-androidTest.apk'
+
+# 63 locales
+locales ['ar', 'ast', 'az', 'bg', 'bn-BD', 'bn-IN', 'ca', 'cs', 
+'cy', 'da', 'de', 'dsb', 'el', 'eo', 'es-AR', 'es-CL', 'es-ES', 'es-MX', 'eu',
+'fa', 'fr', 'fy-NL', 'ga-IE', 'he', 'hi-IN',  'hsb', 'hu', 'hy-AM', 'id', 'it',
+'ja', 'ka', 'kab', 'kk', 'ko', 'lo', 'mix', 'ms', 'my','nb-NO', 'ne-NP', 'nl',
+'nn-NO', 'pl', 'pt-BR', 'ro', 'ru', 'sk', 'sl', 'sq', 'sr', 'sv-SE', 'ta', 'te', 
+'th', 'tr','uk', 'ur','zh-CN', 'zh-TW','en-US']
+
```
Comment (npark-mozilla) [inline Screengrabfile]: So I assume there would be a manual process to check whether the locale lists are up to date? We should clarify the steps in a wiki somewhere.

File: app/src/androidTest/java/org/mozilla/focus/activity/helpers/HostScreencapScreenshotStrategy.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        if (bitmap == null) {
+            bitmap = createDummyScreenShot();
+        }
+
+        screenshotCallback.screenshotCaptured(screenshotName, bitmap);
+    }
+
+    private void takeScreenshotViaHost() {
+        try {
+            final HttpURLConnection connection = (HttpURLConnection) new URL("http://" + HOST_LOOPBACK + ":" + PORT).openConnection();
```
Comment (npark-mozilla) [inline app/src/androidTest/java/org/mozilla/focus/activity/helpers/HostScreencapScreenshotStrategy.java]: This is awesome, just curious, would it make any more reliable if we define a connection at the beginning of the test in an initializer, and close the connection when the object is cleared?  It's just a suggestion

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java [hunk 11]
```diff
@@ -270,8 +271,7 @@ private void takeScreenshotOfNotification(UiDevice device) {
     }
 
     private void takeScreenshotOfEraseSnackbar(UiDevice device) {
-                /* History Erase Notification */
-
+        TestHelper.webView.waitForExists(waitingTime);
```
Comment (npark-mozilla) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: you could also wait for the progressbar to disappear:
 in testhelper:
    static UiObject progressBar = mDevice.findObject(new UiSelector()
            .resourceId("org.mozilla.focus.debug:id/progress")
            .enabled(true));

TestHelper.progressBar.waitForExists(webPageLoadwaitingTime);
Assert.assertTrue(TestHelper.progressBar.waitUntilGone(webPageLoadwaitingTime));

My pending PR has progressbar declaration.

File: Screengrabfile (hunk 9 above)
Comment (pocmo) [reply Screengrabfile]: Yeah, we have multiple lists of locales that I want to make more explicit somewhere:
* A: List of locales we ship in our dev/beta builds
* B: List of locales we ship in releases (B < A)
* C: List of locales we take screenshots of (C = A)

File: app/src/androidTest/java/org/mozilla/focus/activity/helpers/HostScreencapScreenshotStrategy.java (hunk 10 above)
Comment (pocmo) [reply app/src/androidTest/java/org/mozilla/focus/activity/helpers/HostScreencapScreenshotStrategy.java]: Locally it's 100% reliable. So I guess that in automation we just hit the timeouts. But re-using an existing connection might make it faster. :)

**Comment by pocmo** [comment]
In this case I'd actually like to keep them as separate commits. :)

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 1 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 2 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 3 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '\*' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '\*' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 4 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: ',' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 5 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: ',' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 6 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '\+' is not preceded with whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 7 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[PMD\] INFO: This class has too many methods, consider refactoring it\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 8 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[PMD\] INFO: Avoid unused private methods such as 'takeScreenshotOfGooglePlayDialog\(UiDevice\)'\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 1 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 2 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '/' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 3 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '\*' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '\*' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 4 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: ',' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 5 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: ',' is not followed by whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 6 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[Checkstyle\] ERROR: '\+' is not preceded with whitespace\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 7 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[PMD\] INFO: This class has too many methods, consider refactoring it\.

File: app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java (hunk 8 above)
Comment (sputnikci) [inline app/src/androidTest/java/org/mozilla/focus/activity/screenshots/ScreenGrabTest.java]: \[PMD\] INFO: Avoid unused private methods such as 'takeScreenshotOfGooglePlayDialog\(UiDevice\)'\.

**Pull request merged**